include/util/multi_level_partition.hpp

```diff
@@ -0,0 +1,250 @@
+#ifndef OSRM_UTIL_MULTI_LEVEL_PARTITION_HPP
+#define OSRM_UTIL_MULTI_LEVEL_PARTITION_HPP
+
+#include "util/typedefs.hpp"
+#include "util/for_each_pair.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <numeric>
+#include <vector>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+namespace osrm
+{
+namespace util
+{
+namespace detail
+{
+// get the msb of an integer
+// return 0 for integers without msb
+template <typename T> std::size_t highestMSB(T value)
```

Comment:
This can be optimized via lookup tables or SWAR O(lg(N)) as here
https://graphics.stanford.edu/~seander/bithacks.html#IntegerLog
i am for
```


unsigned int v;	         // 32-bit value to find the log2 of 
register unsigned int r; // result of log2(v) will go here
register unsigned int shift;

r =     (v > 0xFFFF) << 4; v >>= r;
shift = (v > 0xFF  ) << 3; v >>= shift; r |= shift;
shift = (v > 0xF   ) << 2; v >>= shift; r |= shift;
shift = (v > 0x3   ) << 1; v >>= shift; r |= shift;
                                        r |= (v >> 1);
```